community/kernel/src/main/java/org/neo4j/kernel/impl/api/cursor/TxAbstractNodeCursor.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.api.cursor;
+
+import org.neo4j.function.Consumer;
+import org.neo4j.graphdb.Direction;
+import org.neo4j.kernel.api.cursor.LabelCursor;
+import org.neo4j.kernel.api.cursor.NodeCursor;
+import org.neo4j.kernel.api.cursor.PropertyCursor;
+import org.neo4j.kernel.api.cursor.RelationshipCursor;
+import org.neo4j.kernel.api.txstate.TransactionState;
+import org.neo4j.kernel.impl.api.state.NodeState;
+
+/**
+ * Overlays transaction state on a {@link NodeCursor}.
+ */
+public abstract class TxAbstractNodeCursor
+        implements NodeCursor
+{
+    protected final TransactionState state;
+    private final Consumer<TxAbstractNodeCursor> cache;
```

Comment:
Type of this field should be `Consumer<? extends TxAbstractNodeCursor>` to avoid casts in child classes like here: https://github.com/neo4j/neo4j/pull/4848/files#diff-417eb484057a4476ab620ff0e9fb1e7cR39.
Same comment applies to `TxAbstractRelationshipCursor`.